### ISSUE TITLE
Degenerate behavior on a micronetwork

### ISSUE THREAD

**Issue by danbst** [C1510/agd_exp#1]
Hello, I've been playing with optimizers and networks and found that AGD converges to `nan` on a particular small network.

The network is  a sole 2x1 matrix of ones, and learns addition of two numbers. Simple, huh? AGD goes great for few epochs, but then `nan`s are thrown into weights.

Here is repro code. Note that ADAM also fails on this example, but SGD works.

```
import math
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.nn.init import orthogonal_
import numpy as np

def singular_value(p):
    sv = math.sqrt(p.shape[0] / p.shape[1])
    if p.dim() == 4:
        sv /= math.sqrt(p.shape[2] * p.shape[3])
    return sv

class AGD:
    @torch.no_grad()
    def __init__(self, net, gain=1.0):

        self.net = net
        self.depth = len(list(net.parameters()))
        self.gain = gain

        for p in net.parameters():
            if p.dim() == 1: raise Exception("Biases are not supported.")
            if p.dim() == 2: 
                orthogonal_(p)
            if p.dim() == 4:
                for kx in range(p.shape[2]):
                    for ky in range(p.shape[3]):
                        orthogonal_(p[:,:,kx,ky])
            p *= singular_value(p)

    @torch.no_grad()
    def step(self):

        G = 0
        for p in self.net.parameters():
            G += singular_value(p) * p.grad.norm(dim=(0,1)).sum()
        G /= self.depth

        log = math.log(0.5 * (1 + math.sqrt(1 + 4*G)))

        for p in self.net.parameters():
            factor = singular_value(p) / p.grad.norm(dim=(0,1), keepdim=True)
            p -= self.gain * log / self.depth * factor * p.grad

        return log

class MyNet(nn.Module):
    def __init__(self, input_size, output_size=1):
        super(MyNet, self).__init__()
        self.w = nn.Linear(input_size, output_size, bias=False)

    def forward(self, x):
        out = self.w(x)
        return out

def get_optim(optim, net, **params):
    if optim == "my":
        return MyNetOptimizer(net, **params)
    elif optim == "adam":
        return torch.optim.Adam(net.parameters(), **params)
    elif optim == "sgd":
        if 'lr' not in params:
            params['lr'] = 0.1
        return torch.optim.SGD(net.parameters(), **params)
    elif optim == "agd":
        return AGD(net, **params)
        
def gradient_explorer(lr=1., std=1., mu=0., optim="my"):
    loss_f = lambda output, target: (output - target).square().mean() / 2
    N_batch = 5
    batch_size = 30
    target_w = torch.ones(2, 1)
    seed = 42
    print("Seed:", seed)
    torch.manual_seed(seed)
    train_data = torch.randn(N_batch, batch_size, target_w.shape[0])*std + mu
    batch_gen = lambda: ((x, x@target_w) for x in train_data)
    
    net = MyNet(input_size=target_w.shape[0])
    optimizer = get_optim(optim, net)

    loss_stat = []
    for epoch in range(50):
        epoch_loss = 0
        if target_w is not None:
            w = list(net.parameters())[0]
            if torch.isclose(w, target_w, rtol=0.0).all():
                break
        for batchx, batchy in batch_gen():
            if len(batchy.shape) == 1:
                batchy = batchy.unsqueeze(0).T
            loss = loss_f(net(batchx), batchy)
            epoch_loss += loss.item()
            loss.backward()
            optimizer.step()
            net.zero_grad()
        loss_stat.append(epoch_loss)
    return net, epoch, loss_stat

net, epoch, loss = gradient_explorer(optim='agd')
print('Learned weighs:', net.w.weight)
print(epoch, loss)
```

The output:
```
Seed: 42
Learned weighs: Parameter containing:
tensor([[nan, nan]], requires_grad=True)
49 [0.1435788195230998, 0.00029988250580004205, 2.881991570946596e-07, 4.060755037890891e-10, 6.149879179294875e-13, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]
```

**Comment by danbst** [comment]
crap, right after posting I found that I messed up with matrix initialization values :man_facepalming: 
It works fine then

**Comment by danbst** [comment]
I was wrong, it probably was a problem with Notebook. The problem is still there, even with correct weight initialization from https://github.com/jxbz/agd/blob/main/agd.py

Still degenerates to nan

```
Seed: 42
Learned weighs: Parameter containing:
tensor([[nan, nan]], requires_grad=True)
49 [3.0006548017263412, 0.05012152074777987, 4.206519566452016e-05, 1.8241266452195815e-08, 1.5179613279916364e-11, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]
```

**Comment by jxbz** [comment]
Hey @danbst, thanks for pointing this out! I just took a look, and the problem seems to be that AGD converges so well that the gradient norm underflows. I could fix the problem by wrapping the AGD weight update in an if statement that checks for underflow:

```python
for p in self.net.parameters():
    if p.grad.norm() > 0:
        factor = singular_value(p) / p.grad.norm(dim=(0,1), keepdim=True)
        p -= self.gain * log / self.depth * factor * p.grad
```
After this, I get:
```bash
Seed: 42
Learned weighs: Parameter containing:
tensor([[1., 1.]], requires_grad=True)
6 [3.0006548315286636, 0.050121524604037404, 4.206507804838111e-05, 1.82418722453892e-08, 1.5586988256631636e-11, 2.5712765927944983e-14]
```

**Comment by jxbz** [comment]
By the way, a minimal code change to fix this issue is to replace the line:
```python
p -= self.gain * log / self.depth * factor * p.grad
```
with:
```python
p -= self.gain * log / self.depth * torch.nan_to_num(factor) * p.grad
```
I've pushed this change into https://github.com/jxbz/agd/blob/main/agd.py. Thanks for pointing it out!
